### PR TITLE
fix: shutdown compactor more quickly on cancel

### PR DESCRIPTION
**Pull request by alamb**
# UPDATE
Whoops: looks like this change got included in https://github.com/influxdata/influxdb_iox/pull/4494 as well accidentally. So I guess this PR now has just a small comment improvement

# Rationale

When I hit Ctrl-C in all in one mode it takes a while (sometimes up to 5 seconds) to shutdown which is annoying

This it because the background loop in `run_compactor` loop sleeps for 5 seconds even if it is cancelled.

# Changes
* Wait for either sleep or timer in compactor loop

# Shutdown before this changes

```shell
******
****** Hit CTRL-C
******
...
2022-05-01T16:15:20.628788Z  INFO ioxd_common: frontend shutdown completed server_type=Ingester
2022-05-01T16:15:20.628870Z  INFO ingester::stream_handler::handler: stream handler shutdown kafka_topic=iox-shared kafka_partition=0
2022-05-01T16:15:20.628887Z  INFO ingester::lifecycle: Lifecycle manager shutdown
2022-05-01T16:15:20.629832Z  INFO influxdb_iox::commands::run::main: done serving, draining futures grpc_bind_address=SocketAddr(127.0.0.1:8081) http_bind_address=Some(SocketAddr(127.0.0.1:8080)) server_type=Router2
2022-05-01T16:15:20.630533Z  INFO ioxd_common: backend shutdown completed server_type=Querier
2022-05-01T16:15:20.630573Z  INFO influxdb_iox::commands::run::main: done serving, draining futures grpc_bind_address=SocketAddr(127.0.0.1:8082) http_bind_address=None server_type=Querier
2022-05-01T16:15:20.631149Z  INFO ioxd_common: backend shutdown completed server_type=Ingester
2022-05-01T16:15:20.631190Z  INFO influxdb_iox::commands::run::main: done serving, draining futures grpc_bind_address=SocketAddr(127.0.0.1:8083) http_bind_address=None server_type=Ingester

******
****** Note the ~4 SECOND  gap in messages here
******
2022-05-01T16:15:24.586060Z  INFO ioxd_common: backend shutdown completed server_type=Compactor
2022-05-01T16:15:24.586209Z  INFO influxdb_iox::commands::run::main: done serving, draining futures grpc_bind_address=SocketAddr(127.0.0.1:8084) http_bind_address=None server_type=Compactor
(arrow_dev) alamb@MacBook-Pro-3:~/Software/influxdb_iox2$
```

# Shutdown with this PR:

(look no gap)

```
******
***** Hit CTRL-C
******
2022-05-01T16:08:08.648679Z  INFO influxdb_iox::commands::run::main: done serving, draining futures grpc_bind_address=SocketAddr(0.0.0.0:8888) http_bind_address=None server_type=Querier
...
2022-05-01T16:08:08.650811Z  INFO influxdb_iox::commands::run::main: done serving, draining futures grpc_bind_address=SocketAddr(127.0.0.1:8084) http_bind_address=None server_type=Compactor
2022-05-01T16:08:08.650842Z  INFO influxdb_iox::commands::run::main: done serving, draining futures grpc_bind_address=SocketAddr(127.0.0.1:8083) http_bind_address=None server_type=Ingester
(arrow_dev) alamb@MacBook-Pro-3:~/Software/influxdb_iox$
```

### CODE REVIEW

File: compactor/src/handler.rs [hunk 1]
```diff
@@ -198,9 +198,13 @@ async fn run_compactor(compactor: Arc<Compactor>, shutdown: CancellationToken) {
...
         let _ = futures::future::join_all(handles).await;
 
-        // if all candidate partitions have been compacted, wait a bit before checking again
+        // if all candidate partitions have been compacted, wait a bit
+        // before checking again, but don'skip sleeping if cancel arrives
         if compactions_run == n_candidates {
-            tokio::time::sleep(Duration::from_secs(5)).await;
+            select! {
+                () = tokio::time::sleep(Duration::from_secs(5)).fuse() => {},
+                () = shutdown.cancelled().fuse() => {}
```
Comment (alamb) [inline compactor/src/handler.rs:206]: the cancel is checked at the start of the loop in this function

File: ioxd_compactor/src/lib.rs [hunk 2]
```diff
@@ -85,7 +85,6 @@ impl<C: CompactorHandler + std::fmt::Debug + 'static> ServerType for CompactorSe
     /// Provide a placeholder gRPC service.
     async fn server_grpc(self: Arc<Self>, builder_input: RpcBuilderInput) -> Result<(), RpcError> {
         let builder = setup_builder!(builder_input, self);
-        // TODO add a service here
```
Comment (alamb) [inline ioxd_compactor/src/lib.rs:88]: drive by cleanup

**Review by carols10cents — APPROVED** [review]

**Comment by alamb** [comment]
FYI @tustvold

**Comment by alamb** [comment]
Whoops: looks like this change got included in https://github.com/influxdata/influxdb_iox/pull/4494 as well accidentally. So I guess this PR now has just a small comment improvement

Updated the PR comments to reflect this, and will merge

**Pull request merged**